source/common/filesystem/filesystem_impl.cc

```diff
@@ -46,96 +34,47 @@ FileSharedPtr InstanceImpl::createFile(const std::string& path, Event::Dispatche
-
-  std::ifstream file(path);
-  if (!file) {
-    throw EnvoyException(fmt::format("unable to read file: {}", path));
-  }
-
-  std::stringstream file_string;
-  file_string << file.rdbuf();
-
-  return file_string.str();
+  return raw_instance_.fileReadToEnd(path);
 }
 
 Api::SysCallStringResult InstanceImpl::canonicalPath(const std::string& path) {
-  // TODO(htuch): When we are using C++17, switch to std::filesystem::canonical.
-  char* resolved_path = ::realpath(path.c_str(), nullptr);
-  if (resolved_path == nullptr) {
-    return {std::string(), errno};
-  }
-  std::string resolved_path_string{resolved_path};
-  ::free(resolved_path);
-  return {resolved_path_string, 0};
+  return raw_instance_.canonicalPath(path);
 }
 
-bool InstanceImpl::illegalPath(const std::string& path) {
-  const Api::SysCallStringResult canonical_path = canonicalPath(path);
-  if (canonical_path.rc_.empty()) {
-    ENVOY_LOG_MISC(debug, "Unable to determine canonical path for {}: {}", path,
-                   ::strerror(canonical_path.errno_));
-    return true;
-  }
-
-  // Platform specific path sanity; we provide a convenience to avoid Envoy
-  // instances poking in bad places. We may have to consider conditioning on
-  // platform in the future, growing these or relaxing some constraints (e.g.
-  // there are valid reasons to go via /proc for file paths).
-  // TODO(htuch): Optimize this as a hash lookup if we grow any further.
-  if (absl::StartsWith(canonical_path.rc_, "/dev") ||
-      absl::StartsWith(canonical_path.rc_, "/sys") ||
-      absl::StartsWith(canonical_path.rc_, "/proc")) {
-    return true;
-  }
-  return false;
-}
+bool InstanceImpl::illegalPath(const std::string& path) { return raw_instance_.illegalPath(path); }
 
-FileImpl::FileImpl(const std::string& path, Event::Dispatcher& dispatcher,
+FileImpl::FileImpl(RawFilePtr&& raw_file, Event::Dispatcher& dispatcher,
                    Thread::BasicLockable& lock, FileSystemStats& stats,
                    std::chrono::milliseconds flush_interval_msec,
                    Thread::ThreadFactory& thread_factory)
-    : path_(path), file_lock_(lock), flush_timer_(dispatcher.createTimer([this]() -> void {
+    : raw_file_(std::move(raw_file)), file_lock_(lock),
+      flush_timer_(dispatcher.createTimer([this]() -> void {
         stats_.flushed_by_timer_.inc();
         flush_event_.notifyOne();
         flush_timer_->enableTimer(flush_interval_msec_);
       })),
-      os_sys_calls_(Api::OsSysCallsSingleton::get()), thread_factory_(thread_factory),
-      flush_interval_msec_(flush_interval_msec), stats_(stats) {
+      thread_factory_(thread_factory), flush_interval_msec_(flush_interval_msec), stats_(stats) {
   open();
 }
 
 void FileImpl::open() {
-  Api::SysCallIntResult result =
-      os_sys_calls_.open(path_, O_RDWR | O_APPEND | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
-  fd_ = result.rc_;
-  if (-1 == fd_) {
+  const auto result = raw_file_->open();
```

Comment:
Explicit type here as it's not obvious from context.

https://google.github.io/styleguide/cppguide.html#auto